### DISCUSSION TITLE
Convert a `Box<T>` into a PyClass

### DISCUSSION THREAD

**Discussion by JackAshwell11** [PyO3/pyo3#2839] [Questions] [answered]
Say if I have this struct:
```rust
#[pyclass]
#[derive(Clone)]
pub struct Leaf {
    left: Option<Box<Leaf>>,
    right: Option<Box<Leaf>>,
    parent: Option<Box<Leaf>>,
}
```
How can I access the `left`, `right`, and `parent` fields from Python since `Leaf` is already a `PyClass`, but it is inside of a `Box`?

**Comment by adamreichold** [comment, marked as answer]
If you want to hand out references to those `Leaf` objects to Python code, store them in the `Py` smart pointer (i.e. on Python's heap with shared mutability using reference counting) instead of using `Box`.

Note that since using `Py` implies shared ownership, you might need to use interior mutability of your `Leaf` structure is not immutable, e.g. using `Option<Py<PyCell<Leaf>>>` which would the Python-compatible equivalent of `Option<Rc<RefCell<Leaf>>`.

**Reply by JackAshwell11** [reply]
I would like those three fields to also be accessible for use in Rust functions so I'm not sure if having the `Py` smart pointer will allow that

**Reply by adamreichold** [reply]
Rust code can certainly also access the content of those `Py` pointers, the only prerequisite is having a GIL token `py: Python` to e.g. call [`Py::as_ref`](https://docs.rs/pyo3/latest/pyo3/prelude/struct.Py.html#method.as_ref), i.e. holding the GIL.

This can either be accomplished by passing that token down the call chain from the Python visible entry points which can always just include a `py: Python` parameter or by explicitly (re-)acquiring the GIL using [`Python::with_gil`](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.with_gil).

**Reply by JackAshwell11** [reply]
So if I'm only going to be accessing those fields a couple times (through some tests in Python), but alot of times in Rust, would this still be the best option?

**Reply by adamreichold** [reply]
It depends, do you need references to those `Leaf` instances that the Python code keeps hold of? If not, you could just clone them in the getter functions and return copies of the relevant subtrees. Otherwise, there is no way around `Py`.

**Reply by JackAshwell11** [reply]
The Python code that would use those attributes would only just check their values and test them, so I shouldn't need any references. But my original idea was a getter, this is my current code:
```rust
#[pymethods]
impl Leaf {
    #[getter(left)]
    fn get_left(&self) -> Option<Leaf> {
        self.left.as_ref().map(|x| Leaf::clone(x))
    }
}
```
Just wasn't sure if it was the best solution.

**Reply by adamreichold** [reply]
> Just wasn't sure if it was the best solution.

If it is only for tests, it should be the simplest solution. Note that you can use [`Option::cloned`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.cloned).

**Reply by JackAshwell11** [reply]
When using `Option::cloned`, I get the error `expected Option<Leaf>, found Cloned<&mut Option<Box<Leaf>>>`. The code:
```rust
    #[getter(left)]
    fn get_left(&self) -> Option<Leaf> {
        self.left.cloned()
    }
```

**Reply by adamreichold** [reply]
The call `Option::as_ref` is still required, i.e. `self.leaf.as_ref().cloned()`.

**Reply by JackAshwell11** [reply]
That still returns a `Option<Box<Leaf>>`

**Reply by adamreichold** [reply]
Of course, the hint on `Option::cloned` was just code golfing to avoid the manual closure. If references are to be shared with Python, `Py<_>` is still the way to go.
